add ioctl() documentation (#146)
phoenix-rtos/phoenix-rtos-doc: libc/functions/i/ioctl.part-impl.md

# Synopsis 
`#include <ioctl.h>`</br>

` int ioctl(int fildes, int request, ... /* arg */);`</br>

## Status
Partially implemented
## Conformance
IEEE Std 1003.1-2017
## Description

The `ioctl()` function shall perform a variety of control functions on devices. For _fildes_ not referring to devices, the functions performed by this call are unspecified.

The _fildes_ argument is an open file descriptor that refers to a device.

The _request_ argument selects the control function to be performed and shall depend on the device being addressed.

The _arg_ argument represents additional information that is needed by this specific device to perform the requested function. The type of _arg_ depends upon the particular control request, but it shall be either an integer or a pointer to a device-specific data structure.

In order to define driver request one should use one of the macros presented below, which are available in the `ioctl.h` header file:

- `_IOC(inout,group,num,len)` - macro which return every possible request 

>   - `inout` - use these defines: IOC_VOID (no data), IOC_OUT (read), IOC_IN (write), IOC_INOUT (read/write)

Reviewer: ```suggestion
  - `inout` - use these defines: `IOC_VOID` (no data), `IOC_OUT` (read), `IOC_IN` (write), `IOC_INOUT` (read/write)
```

>   - `group` - describe group type (collecting the ioctls in groups for a common purpose or a common driver e.g. 'S' for sockets), 8-bit value

Reviewer: ```suggestion
  - `group` - describe group type (collecting the ioctls in groups for a common purpose or a common driver e.g. '`S'` for sockets), 8-bit value
```

>   - `num` - specifie serial number of request, 8 -bit value

Reviewer: ```suggestion
  - `num` - specifies serial number of request, 8 -bit value
```

  - `len` - how big is size of data, 14-bit value

- `_IOV(g,n)` - macro which returns no data transmission request
- `_IOR(g,n,t)` - macro which returns read request
- `_IOW(g,n,t)` - macro which returns write request
- `_IOWR(g,n,t)` - macro which returns read/write request

  - `g` - 8-bit group value
  - `n` - 8-bit serial number
  - `t` - type of data

There are also some checking macros:

- `OCPARM_LEN(x)` - returns data length
- `IOCBASECMD(x)` - returns base command value (2 least significant bytes describing group and serial number)
- `IOCGROUP(x)` - returns group type value

> The resulting requst will be in such form: 2 bits describe direction (`00`: none, `10`: write, `01`: read, `11`: read/write) followed by 14 size bits, followed by an 8-bit group type (collecting the ioctls in groups for a common purpose or a common driver) and an 8-bit serial number.

Reviewer: ```suggestion
The resulting request will be in such form: 2 bits describe direction (`00`: none, `10`: write, `01`: read, `11`: read/write) followed by 14 size bits, followed by an 8-bit group type (collecting the ioctls in groups for a common purpose or a common driver) and an 8-bit serial number.
```

> It is worth to note that to get data send to driver inside message, one should use `ioctl_unapck()`.

Reviewer: ```suggestion
It is worth to note that to get data sent to driver inside message, one should use `ioctl_unapck()`.
```

## Return value

Upon successful completion, `ioctl()` shall return a value other than `-1` that depends on the device control function. Otherwise, it shall return `-1` and set `errno` to indicate the error.

## Errors

The `ioctl()` function may fail if:

 * `EBADF` - The _fildes_ argument is not a valid open file descriptor.

>  * `ENOSYS` - Couldn't get info about calling process.

Reviewer: We rather not use colloquials in the documentation.

```suggestion
 * `ENOSYS` - Couldn't get information about calling process.
```

 * `EINVAL` - Tried to send message to invalid port.

## Example usage

```C
#include <ioctl.h>
#include <fcntl.h>
#include <stdio.h>

> #define REQUEST _IOC(IOC_IN, 'A', 0x01, int)

Reviewer: int?

Author: It is passed to sizeof

# Synopsis 
`#include <ioctl.h>`</br>

` int ioctl(int fildes, int request, ... /* arg */);`</br>

## Status
Partially implemented
## Conformance
IEEE Std 1003.1-2017
## Description

The `ioctl()` function shall perform a variety of control functions on devices. For fildes not referring to devices, the functions performed by this call are unspecified.

The fildes argument is an open file descriptor that refers to a device.

The request argument selects the control function to be performed and shall depend on the device being addressed.

The arg argument represents additional information that is needed by this specific device to perform the requested function. The type of arg depends upon the particular control request, but it shall be either an integer or a pointer to a device-specific data structure.

In order to define driver request one should use below macros and defines:

>     #define IOCPARM_MASK            0x3fff
>     #define IOCPARM_LEN(x)          (((x) >> 16) & IOCPARM_MASK)
>     #define IOCBASECMD(x)           ((x) & ~(IOCPARM_MASK << 16))
>     #define IOCGROUP(x)                     (((x) >> 8) & 0xff)
> 
> 
>     #define IOC_VOID                        0x00000000
>     #define IOC_OUT                         0x40000000
>     #define IOC_IN                          0x80000000
>     #define IOC_INOUT                       (IOC_IN | IOC_OUT)
>     #define IOC_DIRMASK                     0xe0000000
> 
> 
>     _IOC(inout,group,num,len)       ((unsigned long) (inout | ((len & IOCPARM_MASK) << 16) | ((group) << 8) | (num)))
>     _IO(g,n)                        _IOC(IOC_VOID, (g), (n), 0)
>     _IOV(g,n,t)                     _IOC(IOC_VOID, (g), (n), sizeof(t))
>     _IOR(g,n,t)                     _IOC(IOC_OUT, (g), (n), sizeof(t))
>     _IOW(g,n,t)                     _IOC(IOC_IN, (g), (n), sizeof(t))
>     _IOWR(g,n,t)                    _IOC(IOC_INOUT, (g), (n), sizeof(t))

Reviewer: I think we shouldn't present there implementation details, which could potentially change.

```suggestion
 * _IOC(inout,group,num,len) - short description
 * _IOV(g,n) - short description
 * _IOR(g,n,t) 
 * _IOW(g,n,t) 
 * _IOWR(g,n,t) 
```

Above that I would not describe the exact bits, but suggest to use those macros and explain how to use its arguments. For example: One of the following macros shall be passed as _inout_ argument:

 * IOC_VOID - direction not specified
 * IOC_OUT - 
 ...

Group type shall be passed as _group_, typically character are used, for example 'S' for sockets, ... It has to be 8bit long, ...

And please describe this way the remaining arguments, adding that arguments for the rest of macros shall be passed by analogy.

Above that, some simple piece of code with an example could be very helpful. Of course it's only pattern, you don't have to use the same words etc.

Reviewer: Please explain also what is `t` argument.

Reviewer: Please say sth about using `ioctl_unpack()`, which is needed during writing an ioctl handler in a driver.

# Synopsis 
`#include <ioctl.h>`</br>

` int ioctl(int fildes, int request, ... /* arg */);`</br>

## Status
Partially implemented
## Conformance
IEEE Std 1003.1-2017
## Description

The `ioctl()` function shall perform a variety of control functions on devices. For _fildes_ not referring to devices, the functions performed by this call are unspecified.

The _fildes_ argument is an open file descriptor that refers to a device.

The _request_ argument selects the control function to be performed and shall depend on the device being addressed.

The _arg_ argument represents additional information that is needed by this specific device to perform the requested function. The type of _arg_ depends upon the particular control request, but it shall be either an integer or a pointer to a device-specific data structure.

In order to define driver request one should use one of the macros presented below, which are available in the `ioctl.h` header file:

>     _IOC(inout,group,num,len) - macro which return every possible request
> 
> _inout_ - use these defines: IOC_VOID (no data), IOC_OUT (read), IOC_IN (write), IOC_INOUT (read/write)
> _group_ - describe group type (collecting the ioctls in groups for a common purpose or a common driver), 8-bit value
> _num_ - specifie serial number of request, 8 -bit value
> _len_ - how big is size of data, 14-bit value
> 
>     _IOV(g,n) - macro which return no data transmission request
>     _IOR(g,n,t) - macro which return read request
>     _IOW(g,n,t) - macro which return write request
>     _IOWR(g,n,t) - macro which return read/write request

Reviewer: We can firstly list all the possible macros like this and then describe arguments in a similar way:
```suggestion
* `_IOC(inout,group,num,len)` - macro which return every possible request
* `_IOV(g,n)` - macro which returns request without data transmission,
* _IOR(g,n,t) - macro which returns read request,
* _IOW(g,n,t) - macro which returns write request,
* _IOWR(g,n,t) - macro which returns read/write request,

Arguments have been described below:
* _inout_ - use these defines: `IOC_VOID` (no data), IOC_OUT (read), IOC_IN (write), IOC_INOUT (read/write)
* _group_ - describe group type (collecting the ioctls in groups for a common purpose or a common driver), 8-bit value
* _num_ - specifie serial number of request, 8 -bit value
* _len_ - how big is size of data, 14-bit value
```

 - It's always good to use `*` or `-` when you list sth.
 
  - Also remember about highlighting `IOC_VOID` and so on. Suggestion is not fully completed.
  
  - macro which return -> macro which returns

Reviewer: As we discussed you want to indicate what should be passed as _group_ argument (typically a uppercase letter, for example 'S' for sockets <- sth like this)

_g_ - 8-bit group value
_n_ - 8-bit serial number
_t_ - type of data

There are also some checking macros:

    OCPARM_LEN(x) - returns data length
    IOCBASECMD(x) - returns base command value (2 least significant bytes describing group and serial number)
    IOCGROUP(x) - returns group type value

> The resulting requst will be in such form: 2 bits describe direction (00: none, 10: write, 01: read, 11: read/write) followed by 14 size bits, followed by an 8-bit group type (collecting the ioctls in groups for a common purpose or a common driver) and an 8-bit serial number.

Reviewer: ```suggestion
The resulting requst will be in such form: 2 bits describe direction (`00`: none, `10`: write, `01`: read, `11`: read/write) followed by 14 size bits, followed by an 8-bit group type (collecting the ioctls in groups for a common purpose or a common driver) and an 8-bit serial number.
```

It is worth to note that to get data send to driver inside message, one should use `ioctl_unapck()`.

## Return value

> Upon successful completion, `ioctl()` shall return a value other than -1 that depends on the device control function. Otherwise, it shall return -1 and set `errno` to indicate the error.

Reviewer: ```suggestion
Upon successful completion, `ioctl()` shall return a value other than -1 that depends on the device control function. Otherwise, it shall return `-1` and set `errno` to indicate the error.
```
